snuba/utils/profiler.py

```diff
@@ -0,0 +1,63 @@
+import logging
+import socket
+import threading
+import time
+
+import sentry_sdk
+from sentry_sdk.tracing import Transaction
+
+from snuba.state import get_config
+
+logger = logging.getLogger(__name__)
+
+
+def run_ondemand_profiler() -> None:
+    thread = threading.Thread(
+        target=_profiler_main, name="snuba ondemand profiler", daemon=True
+    )
+    thread.start()
+
+
+def _profiler_main() -> None:
+    open_transaction = None
+    own_hostname = socket.gethostname()
+
+    while True:
+        queried_hostnames = get_config("ondemand_profiler_hostnames") or ""
+        queried_hostnames = queried_hostnames.split(",")
+
+        if own_hostname in queried_hostnames and open_transaction is None:
+            # Log an error to Sentry on purpose, if the pod slows down it
+            # should be obvious why.
+            logger.warn("starting ondemand profile for %s", own_hostname)
+
+            with sentry_sdk.Hub.main:
+                open_transaction = sentry_sdk.start_transaction(
+                    name=f"ondemand profile: {own_hostname}", sampled=True
+                )
+                assert isinstance(open_transaction, Transaction)
+                assert open_transaction._profile is not None
+                open_transaction._profile.sampled = True
```

Comment:
Instead of doing this explicitly, you could use the `profiles_sampler` to decide when the profiler is started/sampled. And I assume the reason you call `__enter__` and `__exit__` explicitly below is so you can add this logic before starting the transaction which autoruns the profiler.

Reply:
i have to call `__enter__` and `__exit__` because otherwise the transaction doesn't start, regardless of sampling decision, i think

`profiles_sampler` works, but i feel that i would have to pass some value into the transaction that tells the sampler what to do. i don't want to sample 100% of flask transactions for now

Reply:
Ahh, I see, so the whole thing is controlled by the while loop outside and that either starts a transaction/profile, stops it, or wait 5s.

One caveat to mention here is that the profiler has a 30s limit and will buffer the profile in memory until the transaction ends. So It might be better to start/stop a new transaction every N seconds to flush the profile.

Perhaps something like

```python
while True:
    queried_hostnames = get_config("ondemand_profiler_hostnames") or ""
    queried_hostnames = queried_hostnames.split(",")
    
    if own_hostname in queried_hostnames and open_transaction is None:
       transaction = ...
       # set up the profile
       with transaction:
           time.sleep(5)
    else:
        time.sleep(5)
```

Reply:
>Ahh, I see, so the whole thing is controlled by the while loop outside and that either starts a transaction/profile, stops it, or wait 5s.

yeah I don't even want to profile this thread, i'm interested in the main thread (on _some_ pods)

got it wrt in-memory buffering. i pushed a fix that runs a profile for up to the full 30 seconds before sending the profile to sentry. i think we want to max that time out fully unless we see impact of the overhead in prod and want to turn it off again.

Reply:
There is no way to opt a thread out of being profiled but setting the active thread id should surface whatever is running on the main thread instead. 👍 

I'll let you decide how long to buffer for as long as you're aware that the buffering lasts for the duration of the transaction.

Reply:
i'm not overly concerned about memory usage atm, my biggest fear comes from impact on CPU (or rather, the scheduler hogging the GIL and making the main thread slow). in that sense it would be really nice to have a way to opt out of calling `extract_stack` for certain (or almost all) threads